chain/network/src/peer/rate_counter.rs

```diff
@@ -1,86 +1,93 @@
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
 use std::collections::VecDeque;
-use std::time::{Duration, SystemTime};
-
-const MINUTE: Duration = Duration::from_secs(60);
+use std::time::{Duration, Instant};
 
-/// Stores list of entries for `RateCounter`
-struct Entry {
-    /// bytes since last reset
+/// Represents a single event in time.
+struct Event {
+    /// Time
+    instant: Instant,
+    /// Number of bytes
     bytes: u64,
-    /// Time we created the entry.
-    recorded: SystemTime,
 }
 
-/// A rate counter tracks number of transfers, the amount of data exchanged and the rate of transfer
-/// over the last minute.
-pub struct RateCounter {
-    entries: VecDeque<Entry>,
-    bytes_sum: u64,
+/// Represents all events which happened in last minute.
+#[derive(Default)]
+pub struct TransferStats {
+    /// We keep list of entries not older than 1m.
+    events: VecDeque<Event>,
+    /// Sum of bytes for all entries.
+    bytes_per_min: u64,
 }
 
-impl RateCounter {
-    pub fn new() -> Self {
-        RateCounter { entries: VecDeque::new(), bytes_sum: 0 }
-    }
-
-    /// Increment number of bytes transferred, updating counts and rates.
-    pub fn increment(&mut self, bytes: u64) {
-        let now = SystemTime::now();
-        self.entries.push_back(Entry { bytes, recorded: now });
-        self.bytes_sum += bytes;
-        self.truncate(now);
-    }
+/// Represents cumulative stats per minute.
+#[derive(Eq, PartialEq, Debug)]
+pub struct MinuteStats {
+    /// Byres per minute.
+    pub bytes_per_min: u64,
+    /// Byres per minute.
+    pub count_per_min: usize,
+}
 
-    pub fn bytes_per_min(&self) -> u64 {
-        self.bytes_sum
+impl TransferStats {
+    /// Record event at given `instant` with `bytes` bytes.
+    pub fn record(&mut self, bytes: u64, instant: Instant) {
+        self.bytes_per_min += bytes;
+        self.events.push_back(Event { instant, bytes });
+        self.remove_old_entries(instant);
     }
 
-    pub fn count_per_min(&self) -> u64 {
-        self.entries.len() as u64
+    /// Get stats stored in `MinuteStats` struct.
+    pub fn minute_stats(&mut self, instant: Instant) -> MinuteStats {
+        self.remove_old_entries(instant);
+        MinuteStats { bytes_per_min: self.bytes_per_min, count_per_min: self.events.len() }
     }
 
-    fn truncate(&mut self, now: SystemTime) {
-        // Remove entries older than 1m.
-        while !self.entries.is_empty() && self.entries.front().unwrap().recorded < now - MINUTE {
-            self.bytes_sum -= self.entries.pop_front().unwrap().bytes;
+    /// Remove entries older than 1m.
+    fn remove_old_entries(&mut self, instant: Instant) {
+        while let Some(event) = self.events.pop_front() {
```

Comment:
this assumes that objects in the events queue are in order, right?

Reply:
@mm-near Yes, that's how it works. I updated the documentation to make code easier to read.

Reply:
Yes, I updated docs.